tempodb/backend/gcs/gcs.go

```diff
@@ -186,6 +190,112 @@ func (rw *readerWriter) List(ctx context.Context, keypath backend.KeyPath) ([]st
 	return objects, nil
 }
 
+// ListBlocks implements backend.Reader
+func (rw *readerWriter) ListBlocks(ctx context.Context, keypath backend.KeyPath) (blockIDs []uuid.UUID, compactedBlockIDs []uuid.UUID, err error) {
+	span, ctx := opentracing.StartSpanFromContext(ctx, "readerWriter.ListBlocks")
+	defer span.Finish()
+
+	keypath = backend.KeyPathWithPrefix(keypath, rw.cfg.Prefix)
+	prefix := path.Join(keypath...)
+	if len(prefix) > 0 {
+		prefix += "/"
+	}
+
+	bb := blockboundary.CreateBlockBoundaries(rw.cfg.ListBlocksConcurrency)
+
+	errChan := make(chan error, len(bb))
+	wg := sync.WaitGroup{}
+	mtx := sync.Mutex{}
+
+	var min uuid.UUID
+	var max uuid.UUID
+
+	for i := 0; i < len(bb)-1; i++ {
+		min = uuid.UUID(bb[i])
+		max = uuid.UUID(bb[i+1])
+
+		wg.Add(1)
+		go func(min, max uuid.UUID) {
+			defer wg.Done()
+
+			iter := rw.bucket.Objects(ctx, &storage.Query{
+				Prefix:      prefix,
+				Delimiter:   "",
+				Versions:    false,
+				StartOffset: prefix + min.String(),
+			})
+
+			var parts []string
+			var id uuid.UUID
+
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				default:
+					attrs, err := iter.Next()
+					if errors.Is(err, iterator.Done) {
+						return
+					}
+					if err != nil {
+						errChan <- fmt.Errorf("iterating blocks: %w", err)
+						return
+					}
+
+					parts = strings.Split(attrs.Name, "/")
+					// ie: <tenant>/<blockID>/meta.json
+					if len(parts) != 3 {
```

Comment:
the prefix should contain the tenant id. i would expect the parts to contain 2, but i suppose this is just an a difference between s3 and the other backends? which only return the parts not in the prefix?

Also, wdyt think about these methods failing if not provided with a keypath of length 1? or maybe we should just change the "keypath []string" param to "tenant string"?

Reply:
I think it depends if you include a delimiter or not on s3 as to the number of parts.  The comment there is an example result.

I'm okay with adjusting the interface to be a string instead of a path.

Reply:
not a big deal either way. if you find that changing keypath -> tenantid makes things very difficult you can skip it. it's just closer to the true meaning of what this method is doing and will make it harder to mess up and pass an empty keypath

Reply:
I've adjusted the interface and implementations to use tenant string instead of keypath.